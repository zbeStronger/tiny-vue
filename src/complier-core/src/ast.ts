export const enum NodeTypes {
  INTERPOLATION = "interpolation",
  SIMPLE_EXPRESSION = "simple_expression",
  ELEMENT = "element",
  TEXT = "text",
  ROOT = "root",
}

export const enum TagType {
  START = "start",
  END = "end",
}

export const enum ElementTypes {
  ELEMENT = "element",
}
