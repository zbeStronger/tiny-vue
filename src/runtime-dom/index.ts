import { createRenderer } from "../runtime-core";
//创建元素
function createElement(type) {
  return document.createElement(type);
}
//设置属性
function patchProp(el, key, prevVal, nextVal) {
  const isOn = (key: string) => /^on[A-Z]/.test(key);
  if (isOn(key)) {
    el.setAttribute(key, nextVal);
    const event = key.slice(2).toLowerCase();
    el.addEventListener(event, nextVal);
  } else {
    if (nextVal == null) {
      el.removeAttribute(key);
    } else {
      el.setAttribute(key, nextVal);
    }
  }
}
//插入元素
function insert(el, container, anchor) {
  // container.append(el);
  container.insertBefore(el, anchor || null);
}
function remove(child) {
  const parent = child.parentNode;
  if (parent) {
    parent.removeChild(child);
  }
}
function setTextElement(el, text) {
  el.textContent = text;
}
const renderer: any = createRenderer({
  createElement,
  patchProp,
  insert,
  remove,
  setTextElement,
});

export function createApp(...args) {
  return renderer.createApp(...args);
}

export * from "../runtime-core";
