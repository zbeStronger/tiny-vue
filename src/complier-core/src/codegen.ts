import { NodeTypes } from "./ast";
import { helperNameMap, TO_DISPLAY_STRING } from "./runtimeHelpers";

export function generate(ast) {
  const context: any = createCodegenContext();
  const { push } = context;
  genFunctionPreamble(ast, context);
  const functionName = "render";
  const args = ["_ctx", "_cache"];
  const signature = args.join(", ");
  push(`function ${functionName} (${signature}){`);
  genNode(ast.codegenNode, context);
  push("}");
  return {
    code: context.code,
  };
}

function genFunctionPreamble(ast: any, context) {
  const { push } = context;
  const aliasHelper = (s) => `${helperNameMap[s]} : _${helperNameMap[s]}`;

  if (ast.helpers.length > 0) {
    const VueBinging = "Vue";
    push(`const {${ast.helpers.map(aliasHelper).join(",")}} = ${VueBinging}`);
  }
  push("\n");
  push("return");
}

function genNode(node, context) {
  switch (node.type) {
    case NodeTypes.TEXT:
      getText(node, context);
      break;
    case NodeTypes.INTERPOLATION:
      genInterpolation(node, context);
      break;
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(node, context);
      break;
    default:
      break;
  }
}

function createCodegenContext() {
  const context = {
    code: "",
    helper(key) {
      return `_${helperNameMap[key]}`;
    },
    push(source) {
      context.code += source;
    },
  };
  return context;
}
function getText(node, context) {
  const { push } = context;
  push(`return '${node.content}' `);
}
function genInterpolation(node: any, context: any) {
  const { push, helper } = context;
  push(`${helper(TO_DISPLAY_STRING)}(`);
  genNode(node.content, context);
  push(")");
}
function genExpression(node: any, context: any) {
  const { push } = context;
  push(`${node.content}`);
}
