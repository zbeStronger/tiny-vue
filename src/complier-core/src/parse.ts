import { NodeTypes, TagType } from "./ast";

export function baseParse(content: string) {
  const context = createParserContext(content);
  return creatRoot(parseChildren(context, []));
}
function parseChildren(context, ancestors) {
  const nodes: any[] = [];
  while (!isEnd(context, ancestors)) {
    let node;
    const source = context.source;
    if (source.startsWith("{{")) {
      node = parseInterpolation(context);
    } else if (source[0] === "<") {
      if (/[a-z]/i.test(source[1])) {
        //   console.log("Element");
        node = parseElement(context, ancestors);
      }
    }
    if (!node) {
      node = parseText(context);
    }
    nodes.push(node);
  }

  return nodes;
}

function isEnd(context, ancestors) {
  // 1、source有值
  // 2、遇到结束标签
  const source = context.source;
  if (source.startsWith("</")) {
    for (let i = 0; i < ancestors.length; i++) {
      const tag = ancestors[i].tag;
      if (source.slice(2, 2 + tag.length) === tag) {
        return true;
      }
    }
  }
  // if (parentTag && source.startsWith(`</${parentTag}>`)) {
  //   return true;
  // }

  return !source;
}

function parseInterpolation(context) {
  // {{message}}
  const openDelimiter = "{{";
  const closeDelimiter = "}}";
  const closeIndex = context.source.indexOf(
    closeDelimiter,
    openDelimiter.length
  );
  advanceBy(context, openDelimiter.length);
  const rawContentLength = closeIndex - openDelimiter.length;
  const rawCotent = parseTextData(context, rawContentLength);
  const content = rawCotent.trim();
  advanceBy(context, closeDelimiter.length);
  //   console.log(context);
  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content: content,
    },
  };
}
function createParserContext(content: string) {
  return {
    source: content,
  };
}
function creatRoot(children) {
  return {
    children,
  };
}
function advanceBy(context: any, length: number) {
  context.source = context.source.slice(length);
}
function parseElement(context, ancestors) {
  const element: any = parseTag(context, TagType.START);

  ancestors.push(element);
  const children = parseChildren(context, ancestors);
  ancestors.pop();
  if (startsWithEndTagOpen(context.source, element.tag)) {
    parseTag(context, TagType.END);
  } else {
    throw new Error(`缺失结束标签：${element.tag}`);
  }

  element.children = children;

  return element;
}
function startsWithEndTagOpen(source: string, tag: string) {
  return (
    startsWith(source, "</") &&
    source.slice(2, 2 + tag.length).toLowerCase() === tag.toLowerCase()
  );
}
function startsWith(source: string, searchString: string): boolean {
  return source.startsWith(searchString);
}

function parseTag(context: any, type: TagType) {
  const match: any = /^<\/?([a-z]*)/i.exec(context.source);
  const tag = match[1];
  advanceBy(context, match[0].length);
  advanceBy(context, 1);
  if (type === TagType.END) return;
  return {
    type: NodeTypes.ELEMENT,
    tag: tag,
  };
}

function parseText(context: any) {
  let endIndex = context.source.length;
  let endTokens = ["<", "{{"];
  for (let i = 0; i < endTokens.length; i++) {
    const index = context.source.indexOf(endTokens[i]);
    if (index !== -1 && endIndex > index) {
      endIndex = index;
    }
  }
  const content = parseTextData(context, endIndex);
  return {
    type: NodeTypes.TEXT,
    content,
  };
}
function parseTextData(context, length) {
  const content = context.source.slice(0, length);
  advanceBy(context, content.length);
  return content;
}
