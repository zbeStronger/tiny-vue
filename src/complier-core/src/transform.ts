import { NodeTypes } from "./ast";
import { TO_DISPLAY_STRING } from "./runtimeHelpers";

export function transform(root, options = {}) {
  const context = createTransformContext(root, options);
  traverseNode(root, context);
  createRootCodegen(root);
  root.helpers = [...context.helpers.keys()];
}
function traverseNode(node: any, context) {
  const children = node.children;
  const nodeTransforms = context.nodeTransforms;
  nodeTransforms.forEach((fn) => {
    fn(node);
  });

  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      context.helper(TO_DISPLAY_STRING);
      break;
    case NodeTypes.ROOT:
      traverseChildren(children, context);
      break;
    case NodeTypes.ELEMENT:
      traverseChildren(children, context);
      break;
    default:
      break;
  }
}
function traverseChildren(children, context) {
  for (let i = 0; i < children.length; i++) {
    const node = children[i];
    traverseNode(node, context);
  }
}
function createTransformContext(root: any, options: any) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: new Map(),
    helper(key) {
      context.helpers.set(key, 1);
    },
  };
  return context;
}
function createRootCodegen(root: any) {
  root.codegenNode = root.children[0];
}
