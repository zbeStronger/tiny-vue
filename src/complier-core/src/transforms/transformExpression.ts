import { NodeTypes } from "../ast";

export function transformExpression(node) {
  if (node.type === NodeTypes.INTERPOLATION) {
    node.content = processExpress(node.content);
  }
}
function processExpress(node: any) {
  node.content = `_ctx.${node.content}`;
  return node;
}
