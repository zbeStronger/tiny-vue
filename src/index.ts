export * from "./runtime-dom";
export * from "./reactivity";
import { baseCompile } from "./complier-core/src";
import * as runtimeDom from "./runtime-dom";
import { registerRuntimeCompiler } from "./runtime-dom";
function compileToFunction(template, options = {}) {
  const { code } = baseCompile(template);

  // 调用 compile 得到的代码在给封装到函数内，
  // 这里会依赖 runtimeDom 的一些函数，所以在这里通过参数的形式注入进去
  const render = new Function("Vue", code)(runtimeDom);
  console.log("-----------------", render);
  return render;
}

registerRuntimeCompiler(compileToFunction);
