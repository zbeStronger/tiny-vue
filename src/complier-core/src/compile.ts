import { generate } from "./codegen";
import { baseParse } from "./parse";
import { transform } from "./transform";
import { transformElement } from "./transforms/transformElement";
import { transformExpression } from "./transforms/transformExpression";
import { transfromText } from "./transforms/transformText";

export function baseCompile(template) {
  const ast = baseParse(template);
  transform(ast, {
    nodeTransforms: [transformExpression, transformElement, transfromText],
  });
  return generate(ast);
}
