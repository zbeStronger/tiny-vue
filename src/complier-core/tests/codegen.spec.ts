import { generate } from "../src/codegen";
import { baseParse } from "../src/parse";
import { transform } from "../src/transform";
import { transformElement } from "../src/transforms/transformElement";
import { transformExpression } from "../src/transforms/transformExpression";
import { transfromText } from "../src/transforms/transformText";

describe("codegen", () => {
  it("string", () => {
    const ast = baseParse("hi");
    transform(ast);
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });

  it("interpolation", () => {
    const ast = baseParse("{{msg}}");
    transform(ast, {
      nodeTransforms: [transformExpression],
    });
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });

  it("element", () => {
    const ast = baseParse("<div>hi</div> ");
    transform(ast, {
      nodeTransforms: [transformExpression, transformElement, transfromText],
    });
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });
});
