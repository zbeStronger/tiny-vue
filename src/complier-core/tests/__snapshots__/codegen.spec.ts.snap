// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen element 1`] = `
"
returnfunction render (_ctx, _cache){}"
`;

exports[`codegen interpolation 1`] = `
"const {toDisplayString : _toDisplayString} = Vue
returnfunction render (_ctx, _cache){_toDisplayString(_ctx.msg)}"
`;

exports[`codegen string 1`] = `
"
returnfunction render (_ctx, _cache){return 'hi' }"
`;
